src/storage/src/byte_array.cairo

```diff
@@ -0,0 +1,176 @@
+//! Naive implementation of `Store` trait for `ByteArray`.
+//!
+//! The idea in this implementation is to store the `ByteArray`
+//! data by hashing a key for each element of it.
+//!
+//! 1. First, the `data.len()` value is stored at the given base address.
+//! 2. The `pending_word` is stored computing a key with `PENDING_WORD_KEY`.
+//! 3. The `pending_word_len` is stored computing a key with `PENDING_WORD_LEN_KEY`.
+//! 4. For each element of data, the key is computed using the index of the element in the array.
+//!
+use core::ByteArray;
+use core::hash::LegacyHash;
+use starknet::{
+    storage_read_syscall, storage_write_syscall, storage_base_address_from_felt252,
+    storage_address_to_felt252, storage_address_from_base, storage_address_from_base_and_offset,
+    SyscallResult, SyscallResultTrait, StorageBaseAddress, Store,
+};
+
+const ADDRESS_DOMAIN: u32 = 0;
+const BYTE_ARRAY_KEY: felt252 = '__BYTE_ARRAY__';
+const PENDING_WORD_KEY: felt252 = '__PENDING_WORD__';
+const PENDING_WORD_LEN_KEY: felt252 = '__PENDING_WORD_LEN__';
+
+/// Implementation of `LegacyHash` for `ByteArray`.
+/// The hash is given by the poseidon hash of the `data` and `pending_word` fields.
+impl LegacyHashByteArray of LegacyHash<ByteArray> {
+    fn hash(state: felt252, value: ByteArray) -> felt252 {
+        let mut buffer = array![value.pending_word];
+
+        let mut index = 0;
+
+        loop {
+            if index == value.data.len() {
+                break;
+            }
+
+            buffer.append((*value.data[index]).into());
+
+            index += 1;
+        };
+
+        poseidon::poseidon_hash_span(buffer.span())
+    }
+}
+
+/// Computes a storage address based on the given key.
+///
+/// # Arguments
+///
+/// * `base` - The base storage address (usually given by the name of the field in storage).
+/// * `key` - The custom key to add in a pre-defined array on which poseidon hash is computed.
+fn compute_storage_address(base: StorageBaseAddress, key: felt252) -> StorageBaseAddress {
+    let storage_addr = storage_address_to_felt252(storage_address_from_base(base));
+    let elt_keys: Array<felt252> = array![BYTE_ARRAY_KEY, storage_addr, key];
+
+    storage_base_address_from_felt252(poseidon::poseidon_hash_span(elt_keys.span()))
+}
+
+impl StoreByteArray of Store<ByteArray> {
+    fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<ByteArray> {
+        StoreByteArray::read_at_offset(address_domain, base, 0)
+    }
+
+    fn write(address_domain: u32, base: StorageBaseAddress, value: ByteArray) -> SyscallResult<()> {
+        StoreByteArray::write_at_offset(address_domain, base, 0, value)
+    }
+
+    fn read_at_offset(
```

Comment:
lots of unwraps here can you change those with `?` especially that this function returns a result

Reply:
Fixed in 00e14928a0e72b1a8d58c89422fb7e18515c773f, the one left are in a loop, where Cairo does not support `?`.